src/machine-learning/machine-learning.ts

```diff
@@ -229,68 +232,153 @@ export class MachineLearning implements FirebaseServiceInterface {
-    this.createTime = new Date(model.createTime).toUTCString();
-    this.updateTime = new Date(model.updateTime).toUTCString();
-    if (model.state?.validationError?.message) {
-      this.validationError = model.state?.validationError?.message;
+    this.client = client;
+  }
+
+  get modelId(): string {
+    return extractModelId(this.model.name);
+  }
+
+  get displayName(): string {
+    return this.model.displayName!;
+  }
+
+  get tags(): string[] {
+    return this.model.tags || [];
+  }
+
+  get createTime(): string {
+    return new Date(this.model.createTime).toUTCString();
+  }
+
+  get updateTime(): string {
+    return new Date(this.model.updateTime).toUTCString();
+  }
+
+  get validationError(): string | undefined {
+    return this.model.state?.validationError?.message;
+  }
+
+  get published(): boolean {
+    return this.model.state?.published || false;
+  }
+
+  get etag(): string {
+    return this.model.etag;
+  }
+
+  get modelHash(): string | undefined {
+    return this.model.modelHash;
+  }
+
+  get tfliteModel(): TFLiteModel | undefined {
+    if (this.model.tfliteModel) {
+      return {
+        gcsTfliteUri: this.model.tfliteModel.gcsTfliteUri,
+        sizeBytes: this.model.tfliteModel.sizeBytes,
+      };
+    }
+    return undefined;
+  }
+
+  public equals(obj: Model): boolean {
+    // We only care about the model for equality.
+    return this.model === obj.model;
+  }
+
+  public toJSON(): string {
+    // We can't just return JSON.stringify(this.model)
+    // because it has extra fields and different formats
+    // etc. So we build the expected model object.
+    const jsonModel: {[key: string]: any}  = {
+      modelId: this.modelId,
+      displayName: this.displayName,
+      tags: this.tags,
+      createTime: this.createTime,
+      updateTime: this.updateTime,
+      published: this.published,
+      etag: this.etag,
+    };
+
+    // Also add possibly undefined fields if they exist.
+
+    if (this.validationError) {
+      jsonModel['validationError'] = this.validationError;
     }
-    this.published = model.state?.published || false;
-    this.etag = model.etag;
-    if (model.modelHash) {
-      this.modelHash = model.modelHash;
+
+    if(this.modelHash) {
+      jsonModel['modelHash'] = this.modelHash;
     }
-    if (model.tfliteModel) {
-      this.tfliteModel = {
-        gcsTfliteUri: model.tfliteModel.gcsTfliteUri,
-        sizeBytes: model.tfliteModel.sizeBytes,
-      };
+
+    if (this.tfliteModel) {
+      jsonModel['tfliteModel'] = this.tfliteModel;
     }
+
+    return JSON.stringify(jsonModel);
   }
 
+
+  /**
+   * Locked indicates if there are active long running operations on the model.
+   * Models may not be modified when they are locked.
+   */
   public get locked(): boolean {
-    // Backend does not currently return locked models.
-    // This will likely change in future.
-    return false;
+    return (this.model.activeOperations?.length || 0) > 0;
```

Comment:
`?? 0` is safer than `|| 0` in most circumstances.

Reply:
done